File: apps/api/aqapi/quest/entity/quest_members_entity.py
from sqlalchemy import Column, Integer, ForeignKey, DateTime, CheckConstraint, UniqueConstraint
from sqlalchemy.orm import relationship, Mapped, mapped_column
from sqlalchemy.sql import func
from aqapi.core.entity.base_entity import BaseEntity
from aqapi.core.entity.base_history_entity import BaseHistoryEntity
from aqapi.core.entity.base_translation_entity import BaseTranslationEntity
from aqapi.core.config.db_config import DB_CONF


class QuestMembersEntity(BaseEntity):
    """クエストメンバーエンティティ
    
    家族クエストに参加している子供の情報を管理するエンティティです。
    """

    __tablename__ = "quest_members"
    __table_args__ = (
        # レベルは0より大きい
        CheckConstraint("current_level > 0", name="chk_quest_members_current_level_positive"),
        # 達成日時は公開日時以降
        CheckConstraint("(achieved_at IS NULL) OR (achieved_at IS NOT NULL AND achieved_at >= published_at)", name="chk_quest_members_achieved_after_published"),
        # 一意制約
        UniqueConstraint("family_quest_id", "member_id"),
    )

    family_quest_id: Mapped[int] = mapped_column(Integer, ForeignKey("family_quests.id", ondelete="CASCADE"), nullable=False, comment="家族クエストID")
    member_id: Mapped[int] = mapped_column(Integer, ForeignKey("children.id", ondelete="CASCADE"), nullable=False, comment="子供ID")
    current_level: Mapped[int] = mapped_column(Integer, nullable=False, default=1, comment="現在のレベル")
    status_id: Mapped[int] = mapped_column(Integer, ForeignKey("quest_member_statuses.id", ondelete="RESTRICT"), nullable=False, comment="クエストステータスID")
    published_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now(), comment="クエスト公開日時")
    achieved_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=True, comment="クエスト達成日時")

    family_quest = relationship("FamilyQuestsEntity", foreign_keys=[family_quest_id])
    member = relationship("ChildrenEntity", foreign_keys=[member_id])
    status = relationship("QuestMemberStatusesEntity", foreign_keys=[status_id])

class QuestMembersHistoryEntity(BaseHistoryEntity):
    """クエストメンバー履歴エンティティ"""

    __tablename__ = "quest_members_history"

    family_quest_id: Mapped[int] = mapped_column(Integer)
    member_id: Mapped[int] = mapped_column(Integer)
    current_level: Mapped[int] = mapped_column(Integer)
    status_id: Mapped[int] = mapped_column(Integer)
    published_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True))
    achieved_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True))

    @classmethod
    def from_source(cls, source: "QuestMembersEntity"):
        """元のレコードから履歴エンティティを生成"""
        return cls(
            source_id=source.id,
            family_quest_id=source.family_quest_id,
            member_id=source.member_id,
            current_level=source.current_level,
            status_id=source.status_id,
            published_at=source.published_at,
            achieved_at=source.achieved_at,
            source_created_at=source.created_at,
            source_updated_at=source.updated_at,
        )

File: packages/backend/src/features/quest/entity/questMembersEntity.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToOne,
  JoinColumn,
  Check,
  Unique,
} from "typeorm";
import { BaseTransactionEntity } from "../../../core/entity/baseTransactionEntity";
import { BaseHistoryEntity } from "../../../core/entity/baseHistoryEntity";
import { ChildEntity } from "../../child/entity/childEntity";
import { QuestMemberStatusEntity } from "./questMemberStatusEntity";

/**
 * クエストメンバーエンティティ
 * 
 * 家族クエストに参加している子供の情報を管理するエンティティです。
 */
@Entity("quest_members")
@Unique("uq_quest_members", ["family_quest_id", "member_id"])
@Check("chk_quest_members_current_level_positive", "current_level > 0")
@Check("chk_quest_members_achieved_after_published", "(achieved_at IS NULL) OR (achieved_at IS NOT NULL AND achieved_at >= published_at)")
export class QuestMembersEntity extends BaseTransactionEntity {
  @Column({ type: "int", nullable: false, comment: "家族クエストID" })
  family_quest_id!: number;
  @Column({ type: "int", nullable: false, comment: "子供ID" })
  member_id!: number;
  @Column({ type: "int", nullable: false, default: 1, comment: "現在のレベル" })
  current_level!: number;
  @Column({ type: "int", nullable: false, comment: "クエストステータスID" })
  status_id!: number;
  @Column({ type: "timestamp", nullable: false, default: () => "CURRENT_TIMESTAMP", comment: "クエスト公開日時" })
  published_at!: Date;
  @Column({ type: "timestamp", nullable: true, comment: "クエスト達成日時" })
  achieved_at!: Date | null;

  @ManyToOne("FamilyQuestEntity", { nullable: false, onDelete: "CASCADE" })
  @JoinColumn({ name: "family_quest_id", referencedColumnName: "id", foreignKeyConstraintName: "fk_quest_members_family_quest_id" })
  family_quest!: any;
  @ManyToOne(() => ChildEntity, { nullable: false, onDelete: "CASCADE" })
  @JoinColumn({ name: "member_id", referencedColumnName: "id", foreignKeyConstraintName: "fk_quest_members_member_id" })
  member!: ChildEntity;
  @ManyToOne(() => QuestMemberStatusEntity, { nullable: false, onDelete: "RESTRICT" })
  @JoinColumn({ name: "status_id", referencedColumnName: "id", foreignKeyConstraintName: "fk_quest_members_status_id" })
  status!: QuestMemberStatusEntity;
}

/**
 * クエストメンバー履歴エンティティ
 */
@Entity("quest_members_history")
export class QuestMembersHistoryEntity extends BaseHistoryEntity {
  @Column({ type: "int" })
  family_quest_id!: number;
  @Column({ type: "int" })
  member_id!: number;
  @Column({ type: "int" })
  current_level!: number;
  @Column({ type: "int" })
  status_id!: number;
  @Column({ type: "timestamp" })
  published_at!: Date;
  @Column({ type: "timestamp" })
  achieved_at!: Date | null;

  /**
   * サブクラス固有の属性をセット
   */
  protected static setSpecificAttrs(instance: QuestMembersHistoryEntity, source: QuestMembersEntity): void {
    instance.family_quest_id = source.family_quest_id;
    instance.member_id = source.member_id;
    instance.current_level = source.current_level;
    instance.status_id = source.status_id;
    instance.published_at = source.published_at;
    instance.achieved_at = source.achieved_at;
  }
}
